File: 112.path-sum.cpp
/*
 * @lc app=leetcode id=112 lang=cpp
 *
 * [112] Path Sum
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution
{
public:
    bool hasPathSum_rec(TreeNode *root, int targetSum, int cur_sum)
    {
        // basecase
        if (!root)
        {
            return false;
        }

        if (cur_sum + root->val == targetSum && !root->left && !root->right)
        {
            return true;
        }

        return hasPathSum_rec(root->left, targetSum, cur_sum + root->val) || hasPathSum_rec(root->right, targetSum, cur_sum + root->val);
    }

    bool hasPathSum(TreeNode *root, int targetSum)
    {
        return hasPathSum_rec(root, targetSum, 0);
    }
};
// @lc code=end

File: 112.path-sum.py
#
# @lc app=leetcode id=112 lang=python3
#
# [112] Path Sum
#

# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def hasPathSum(self, root: Optional[TreeNode], targetSum: int) -> bool:

        # lets use a dfs
        def dfs(root, curr_sum):
            if root == None:
                return False
            
            curr_sum += root.val
            if not root.left and not root.right:
                return curr_sum == targetSum

            return any([dfs(root.left, curr_sum),dfs(root.right, curr_sum)])
        
        return  dfs(root, 0)



        
# @lc code=end

